### PR TITLE
FIFO attempt

### PR DESCRIPTION
**Pull request by Rinnerdos**
This attempt doesn't implement FIFO yet, but I hope that I created the first steps to be able to do it. To me it seems that now only the most perished products need to be passed to the filter, which should be relatively easy with the duplicates objects which gives the duplicates in a container, and their index. I might be wrong of course, and it also might be slightly redundant/too complex. Also, the items can be spread over multiple containers, and I'm not sure if that is easy to incorprate in my method.

### CODE REVIEW

File: src/models/classes/container.js [hunk 1]
```diff
@@ -38,6 +38,44 @@ export default class Container extends StorageCore
...
+               perishTimeArray.push(this.items[i].values.perishable);
+               perishNameArray.push(this.items[i].name);
+           }
+        }
+
+        // https://stackoverflow.com/a/18417854
+        // Creates an object with the duplicate name,
+        // and their index in the original array.
+
+        Array.prototype.getDuplicates = function () {
```
Comment (N-Wouda) [inline src/models/classes/container.js]: Not really problematic, but don't do this here. You're extending the Array prototype in a place where this is _not_ expected. Better to do this in main.js, by including a new file there called e.g. `extensions.js`

File: src/models/classes/container.js [hunk 2]
```diff
@@ -38,6 +38,44 @@ export default class Container extends StorageCore
...
+            let duplicates = {};
+            for (var i = 0; i < this.length; i++) {
```
Comment (N-Wouda) [inline src/models/classes/container.js]: `let`.

File: src/models/classes/container.js [hunk 3]
```diff
@@ -38,6 +38,44 @@ export default class Container extends StorageCore
      */
     removeItem(product)
     {
+        // Maken an array of the perish time and corresponding product name, which can be compared later.
+        let perishTimeArray = [];
+        let perishNameArray = [];
+        let perishDuplicates = {};
+        for (let i = 0; i < this.items.length; i++) {
```
Comment (N-Wouda) [inline src/models/classes/container.js]: `this.items.forEach((product) => {...})`?

**Comment by N-Wouda** [comment]
Attempt to align/structure your code a bit better. Not just in terms of _what_ you wrote, but also _how_ it looks. This is a bit messy.

**Comment by N-Wouda** [comment]
How is this going?

**Comment by Rinnerdos** [comment]
See commits

File: src/models/classes/container.js [hunk 4]
```diff
@@ -36,19 +36,28 @@ export default class Container extends StorageCore
...
+    removeItem(product, perishedQuantity)
     {
         this.items = this.items.filter(function (item) {
             if (product.name != item.name) {
                 return item;
             }
 
-            const removedQuantity = Math.min(product.values.quantity, item.values.quantity);
+            if (product.values.isPerishable === true) {
+                const removedQuantity = Math.min(product.values.quantity, perishedQuantity)
```
Comment (N-Wouda) [inline src/models/classes/container.js]: Product is the external value, perishedQuantity is.. too? You need to check against `item`, which is the product actually in the container.

File: src/models/classes/container.js [hunk 5]
```diff
@@ -36,19 +36,28 @@ export default class Container extends StorageCore
...
-    removeItem(product)
+    removeItem(product, perishedQuantity)
     {
         this.items = this.items.filter(function (item) {
             if (product.name != item.name) {
                 return item;
             }
 
-            const removedQuantity = Math.min(product.values.quantity, item.values.quantity);
+            if (product.values.isPerishable === true) {
```
Comment (N-Wouda) [inline src/models/classes/container.js]:
```javascript
if (product.values.isPerishable) {
```

Comment (Rinnerdos) [reply src/models/classes/container.js]: Ah yeah, forgot about that :D

File: src/models/classes/container.js [hunk 6]
```diff
@@ -36,16 +36,16 @@ export default class Container extends StorageCore
...
         this.items = this.items.filter(function (item) {
             if (product.name != item.name) {
                 return item;
             }
 
-            if (product.values.isPerishable === true) {
-                const removedQuantity = Math.min(product.values.quantity, perishedQuantity)
-             
+            if (product.values.isPerishable) {
+                const removedQuantity = perishedProduct.values.quantity;
```
Comment (N-Wouda) [inline src/models/classes/container.js]: `Math.min(product.values.quantity, item.values.quantity, perishedQuantity);`

You can only delete the bare minimum of this (but you should not use `perishQuantity`, but rather update the value on `product` outside this function).

Comment (Rinnerdos) [reply src/models/classes/container.js]: Yeah that was dumb of me. I now made it 

```Math.min(product.values.quantity, perishedProduct.values.quantity);```

But I just keep struggling with that it doesn't stop removing from a container after the first object has been removed (when I want it to).

File: src/models/classes/container.js [hunk 7]
```diff
@@ -36,19 +36,28 @@ export default class Container extends StorageCore
...
 
-            product.values.quantity -= removedQuantity;
-            item.values.quantity -= removedQuantity;
+                product.values.quantity -= removedQuantity;
+                item.values.quantity -= removedQuantity;
 
-            return item.values.quantity;
+                return item.values.quantity;
```
Comment (N-Wouda) [inline src/models/classes/container.js]: You should return `product.values.quantity`

Don't return here, use the one at the end of the function.

**Pull request merged**